File: Daily/1671.minimum-number-of-removals-to-make-mountain-array.cpp
/*
 * @lc app=leetcode id=1671 lang=cpp
 *
 * [1671] Minimum Number of Removals to Make Mountain Array
 */

#include <vector>

// @lc code=start
class Solution {
 public:
  int minimumMountainRemovals(const std::vector<int>& nums) {
    const int kNumsSize = static_cast<int>(nums.size());
    std::vector<int> lisFromLeft(kNumsSize, 1);
    std::vector<int> lisFromRight(kNumsSize, 1);

    // LIS from left
    for (int i = 0; i < kNumsSize; i++) {
      for (int j = i - 1; j >= 0; j--) {
        if (nums[j] < nums[i]) {
          lisFromLeft[i] = std::max(lisFromLeft[i], lisFromLeft[j] + 1);
        }
      }
    }

    // LIS from right
    for (int i = kNumsSize - 1; i >= 0; i--) {
      for (int j = i + 1; j < kNumsSize; j++) {
        if (nums[i] > nums[j]) {
          lisFromRight[i] = std::max(lisFromRight[i], lisFromRight[j] + 1);
        }
      }
    }

    int maxMountainSize = 0;
    for (int i = 0; i < kNumsSize; i++) {
      const int fromLeftLisSize = lisFromLeft[i];
      const int fromRightLisSize = lisFromRight[i];
      if (fromLeftLisSize < 2 || fromRightLisSize < 2) {
        continue;
      }
      maxMountainSize =
          std::max(maxMountainSize, fromLeftLisSize + fromRightLisSize - 1);
    }

    return kNumsSize - maxMountainSize;
  }
};
// @lc code=end

int main() {
  std::vector<int> nums = {2, 1, 1, 5, 6, 2, 3, 1};
  Solution sol;
  sol.minimumMountainRemovals(nums);
}

File: src/2022-intern/1671.minimum-number-of-removals-to-make-mountain-array.js
/*
 * @lc app=leetcode id=1671 lang=javascript
 *
 * [1671] Minimum Number of Removals to Make Mountain Array
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number}
 */
var minimumMountainRemovals = function (nums) {
  /**
   * increase[i] 以 i-1 为结尾，最长的递增序列长度
   * decrease[i] 以 i+1 为开头，最长的递减序列长度
   */

  const n = nums.length;
  const increase = new Array(n);
  const decrease = new Array(n);
  increase.fill(0);
  decrease.fill(0);

  for (let i = 0; i < n; i++) {
    for (let j = i - 1; j >= 0; j--) {
      if (nums[j] < nums[i]) {
        increase[i] = Math.max(increase[i], increase[j] + 1);
      }
    }
  }

  for (let i = n - 1; i >= 0; i--) {
    for (let j = i + 1; j < n; j++) {
      if (nums[j] < nums[i]) {
        decrease[i] = Math.max(decrease[i], decrease[j] + 1);
      }
    }
  }

  let maxLength = -Infinity;
  for (let i = 1; i < n - 1; i++) {
    // 对于 nums[i]，两侧的数字必须都存在才行
    if (increase[i] !== 0 && decrease[i] !== 0) {
      maxLength = Math.max(increase[i] + decrease[i] + 1, maxLength);
    }
  }

  return n - maxLength;
};
// @lc code=end
